File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();

        vector<vector<int>> ans;

        sort(nums.begin(), nums.end());

        for (int i=0; i<n; i++) {
            if (i > 0 && nums[i] == nums[i-1]) continue;

            int l = i+1, r = n-1;
            while (l < r) {
                int sm = nums[i] + nums[l] + nums[r];
                if (sm > 0) {
                    r--;
                } else if (sm < 0) {
                    l++;
                } else {
                    ans.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    while (l < r && nums[l] == nums[l-1]) {
                        l++;
                    }
                }
            }
        }
        return ans;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        Set<List<Integer>> st = new HashSet<>();
        List<Integer> v = new ArrayList<>();
        dfs(target, nums, v, st);

        List<List<Integer>> ans = new ArrayList<>();
        for (List<Integer> l: st) {
            ans.add(l);
        }
        return ans;
    }

    void dfs(int target, int[] nums, List<Integer> v, Set<List<Integer>> st) {
        if (target == 0) {
            List<Integer> temp = new ArrayList<>(v);
            Collections.sort(temp);
            st.add(temp);
            return;
        }

        for (int x: nums) {
            if (target >= x) {
                v.add(x);
                dfs(target - x, nums, v, st);
                v.removeLast();
            }
        }
    }
}
